### ISSUE TITLE
Need @JsonScalaEnumeration to determine key type when it is present (only for full test suite run)

### ISSUE THREAD

**Issue by kgignatyev** [FasterXML/jackson-module-scala#182]
I have tried 2.5.0 version of everything 
 jackson-annotation.version>2.5.0</jackson-annotation.version>
        jackson-core.version>2.5.0</jackson-core.version>
        jackson-databind.version>2.5.0</jackson-databind.version>
        jackson-module-scala.version>2.5.0</jackson-module-scala.version>
      
and encounter strange behavior (not existent if I set versions to 2.4.4):
When I run all tests for my project a lot of them reliably fail with (not random, same tests fail every time)
 Cause: com.fasterxml.jackson.databind.JsonMappingException: Need @JsonScalaEnumeration to determine key type
 
However, if I run any single of those test with 
mvn test -Dsuites=test class name here
then test passes.....
I tried to creates test to reproduce issue in isolation, but was not able, it can deserialize just fine. 
Apologize for not being able to reproduce, just in case you might have some ideas below are bits of relevant information:

Here are bits of information:
class MotoLocation  extends Serializable with Identifiable
with GeoLocation with Address with McyQuantities  with Auditable with AverageRating with Owned with SEO{.....

trait McyQuantities {

  @JsonScalaEnumeration(classOf[MotorcycleClassType])
  @BeanProperty var mcyTypesCount:mutable.HashMap[MotorcycleClass,Int] = mutable.HashMap()
  def totalMotorcycles(): Int = mcyTypesCount.foldLeft(0)((r,entry)=>r+entry._2)

}

class MotorcycleClassType extends TypeReference[MotorcycleClass.type ]

object MotorcycleClass extends Enumeration{
  type MotorcycleClass = Value
  val atv ,
  cruiser ,
  dualsport ,
  minibike ,
  motocross ,
  other ,
  prototype ,
  speedway ,
  standard ,
  supermotard ,
  touring ,
  trial ,
  trike ,
  scooter ,
  sport  = Value
}

collections configured as JacksonDBCollection.wrap(db.getCollection(simpleName), typeClass, keyType, org.mongojack.internal.MongoJackModule.configure(McyObjectMapper.mongoSideMapper)

there are 2 mappers used in the application

  val mapper: ObjectMapper = {
    val res = new ObjectMapper
    res.registerModule(DefaultScalaModule)
    res.setDateFormat(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss"))
    res
  }

  val mongoSideMapper: ObjectMapper = {
      val res = new ObjectMapper
      res.registerModule(DefaultScalaModule)
      res
  }

CompanyServiceTest:
Exception encountered when invoking run on a nested suite - Read operation to server localhost:27017 failed on database mcytravel *** ABORTED *** (186 milliseconds)
  com.mongodb.MongoException$Network: Read operation to server localhost:27017 failed on database mcytravel
  at com.mongodb.DBTCPConnector.innerCall(DBTCPConnector.java:297)
  at com.mongodb.DBTCPConnector.call(DBTCPConnector.java:268)
  at com.mongodb.DBCollectionImpl.find(DBCollectionImpl.java:84)
  at com.mongodb.DBCollectionImpl.find(DBCollectionImpl.java:66)
  at com.mongodb.DBCursor._check(DBCursor.java:498)
  at com.mongodb.DBCursor._hasNext(DBCursor.java:621)
  at com.mongodb.DBCursor.hasNext(DBCursor.java:657)
  at org.mongojack.DBCursor.hasNext(DBCursor.java:348)
  at org.mongojack.JacksonDBCollection.findOne(JacksonDBCollection.java:1364)
  at org.mongojack.JacksonDBCollection.findOne(JacksonDBCollection.java:1334)
  at org.mongojack.JacksonDBCollection.findOne(JacksonDBCollection.java:1308)
  at kgi.mcytravel.wsite.services.MotoLocationsService.ensureSEOid(MotoLocationsService.scala:139)
  at kgi.mcytravel.wsite.services.MotoLocationsService.save(MotoLocationsService.scala:147)
  at kgi.mcytravel.wsite.services.TestHelperService.createRandomMotoLocation(TestHelperService.scala:99)
  at kgi.mcytravel.wsite.services.McyTravelTester.beforeAll(McyTravelTester.scala:47)
  at org.scalatest.BeforeAndAfterAll$class.beforeAll(BeforeAndAfterAll.scala:187)
  at kgi.mcytravel.wsite.services.McyTravelTester.beforeAll(McyTravelTester.scala:17)
  at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:253)
  at kgi.mcytravel.wsite.services.McyTravelTester.run(McyTravelTester.scala:17)
  at org.scalatest.Suite$class.callExecuteOnSuite$1(Suite.scala:1492)
  at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1528)
  at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1526)
  at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
  at org.scalatest.Suite$class.runNestedSuites(Suite.scala:1526)
  at org.scalatest.tools.DiscoverySuite.runNestedSuites(DiscoverySuite.scala:29)
  at org.scalatest.Suite$class.run(Suite.scala:1421)
  at org.scalatest.tools.DiscoverySuite.run(DiscoverySuite.scala:29)
  at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:55)
  at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2563)
  at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2557)
  at scala.collection.immutable.List.foreach(List.scala:381)
  at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:2557)
  at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1044)
  at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1043)
  at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:2722)
  at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1043)
  at org.scalatest.tools.Runner$.main(Runner.scala:860)
  at org.scalatest.tools.Runner.main(Runner.scala)
  Cause: com.fasterxml.jackson.databind.JsonMappingException: Need @JsonScalaEnumeration to determine key type
 at [Source: de.undercouch.bson4jackson.io.LittleEndianInputStream@50119639; pos: 839] (through reference chain: kgi.mcytravel.wsite.data.db.MotoLocation["mcyTypesCount"])
  at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:148)
  at com.fasterxml.jackson.databind.DeserializationContext.mappingException(DeserializationContext.java:843)
  at com.fasterxml.jackson.module.scala.deser.EnumerationKeyDeserializer.deserializeKey(EnumerationDeserializerModule.scala:85)
  at com.fasterxml.jackson.module.scala.deser.EnumerationKeyDeserializer.deserializeKey(EnumerationDeserializerModule.scala:73)
  at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBind(MapDeserializer.java:404)
  at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:333)
  at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:76)
  at com.fasterxml.jackson.module.scala.deser.UnsortedMapDeserializer.deserialize(UnsortedMapDeserializerModule.scala:39)
  at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:523)
  at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:95)
  at com.fasterxml.jackson.databind.deser.impl.BeanPropertyMap._findDeserializeAndSet2(BeanPropertyMap.java:313)
  at com.fasterxml.jackson.databind.deser.impl.BeanPropertyMap.findDeserializeAndSet(BeanPropertyMap.java:291)
  at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:248)
  at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:136)
  at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3532)
  at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:1934)
  at org.mongojack.internal.stream.JacksonDBDecoder.decode(JacksonDBDecoder.java:77)
  at com.mongodb.Response.<init>(Response.java:85)
  at com.mongodb.DBPort$1.execute(DBPort.java:153)
  at com.mongodb.DBPort$1.execute(DBPort.java:147)
  at com.mongodb.DBPort.doOperation(DBPort.java:176)
  at com.mongodb.DBPort.call(DBPort.java:147)
  at com.mongodb.DBTCPConnector.innerCall(DBTCPConnector.java:289)
  at com.mongodb.DBTCPConnector.call(DBTCPConnector.java:268)
  at com.mongodb.DBCollectionImpl.find(DBCollectionImpl.java:84)
  at com.mongodb.DBCollectionImpl.find(DBCollectionImpl.java:66)
  at com.mongodb.DBCursor._check(DBCursor.java:498)
  at com.mongodb.DBCursor._hasNext(DBCursor.java:621)
  at com.mongodb.DBCursor.hasNext(DBCursor.java:657)
  at org.mongojack.DBCursor.hasNext(DBCursor.java:348)
  at org.mongojack.JacksonDBCollection.findOne(JacksonDBCollection.java:1364)
  at org.mongojack.JacksonDBCollection.findOne(JacksonDBCollection.java:1334)
  at org.mongojack.JacksonDBCollection.findOne(JacksonDBCollection.java:1308)
  at kgi.mcytravel.wsite.services.MotoLocationsService.ensureSEOid(MotoLocationsService.scala:139)
  at kgi.mcytravel.wsite.services.MotoLocationsService.save(MotoLocationsService.scala:147)
  at kgi.mcytravel.wsite.services.TestHelperService.createRandomMotoLocation(TestHelperService.scala:99)
  at kgi.mcytravel.wsite.services.McyTravelTester.beforeAll(McyTravelTester.scala:47)
  at org.scalatest.BeforeAndAfterAll$class.beforeAll(BeforeAndAfterAll.scala:187)
  at kgi.mcytravel.wsite.services.McyTravelTester.beforeAll(McyTravelTester.scala:17)
  at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:253)
  at kgi.mcytravel.wsite.services.McyTravelTester.run(McyTravelTester.scala:17)
  at org.scalatest.Suite$class.callExecuteOnSuite$1(Suite.scala:1492)
  at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1528)
  at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1526)
  at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
  at org.scalatest.Suite$class.runNestedSuites(Suite.scala:1526)
  at org.scalatest.tools.DiscoverySuite.runNestedSuites(DiscoverySuite.scala:29)
  at org.scalatest.Suite$class.run(Suite.scala:1421)
  at org.scalatest.tools.DiscoverySuite.run(DiscoverySuite.scala:29)
  at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:55)
  at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2563)
  at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2557)
  at scala.collection.immutable.List.foreach(List.scala:381)
  at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:2557)
  at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1044)
  at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1043)
  at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:2722)
  at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1043)
  at org.scalatest.tools.Runner$.main(Runner.scala:860)
  at org.scalatest.tools.Runner.main(Runner.scala)
RefDataServiceTest:

**Comment by christophercurrie** [comment]
Thanks for the report. I'm not 100% sure, but I suspect it might be in the interaction with `JacksonScalaEnumeration`, `BeanProperty`, and updates to the property detection logic in 2.5.

If my theory is correct, then a workaround to your issue would be to replace 

```
@JacksonScalaEnumeration(...)
```

with 

```
import scala.annotation.meta._

@(JacksonScalaEnumeration @param @getter @beanGetter)(...)
```

to force the annotation to appear on all the places that the Scala module might potentially be looking for it.

Let me know if that works for you, and I'll try to develop a test for this, and get a fix out for 2.5.1.

**Comment by kgignatyev** [comment]
Thank you for looking. That suggestion does not fix the problem. 
What puzzles me is that annotation and everything works fine when tested alone or when tests which are failing in the group are executed individually.
This is how tests are configured in the project:
 
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>${plugin.surefire}</version>
                <executions>
                    <execution>
                        <id>default-test</id>
                        
                        <phase>none</phase>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.scalatest</groupId>
                <artifactId>scalatest-maven-plugin</artifactId>
                <version>${plugin.scalatest}</version>
                <configuration>
                    <reportsDirectory>${project.build.directory}/surefire-reports</reportsDirectory>
                    <stdout>W</stdout>

                    <junitxml>.</junitxml>
                    <filereports>WDF TestSuite.txt</filereports>
                </configuration>
                <executions>
                    <execution>
                        <id>scala-test</id>
                        <goals>
                            <goal>test</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>

**Comment by christophercurrie** [comment]
It looks like this might actually be related to running the code under Java 7/8; those versions do not guarantee the order that reflection discovers properties. Java 6 didn't officially, either, but in practice it was always discovered in declaration order. I'm able to reproduce the issue with a local test, but not all the time, it varies from run to run.

Can you confirm you're running your tests on Java 7 or 8? I'll see what can be done to make the property detection more reliable.

**Comment by kgignatyev** [comment]
Yes, running tests under java 8
java version "1.8.0_31"
Java(TM) SE Runtime Environment (build 1.8.0_31-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.31-b07, mixed mode)
Darwin kgi-mbp-i7.lan 12.5.0 Darwin Kernel Version 12.5.0: Sun Sep 29 13:33:47 PDT 2013; root:xnu-2050.48.12~1/RELEASE_X86_64 x86_64
OSX 10.8.5

**Comment by christophercurrie** [comment]
Ok, looks like the changes in 2.5.0 introduced a bug when running under Java 7+ and its non-deterministic reflection discovery order. Fix will go out with 2.5.1. In the mean time you can use 2.4.5.

**Comment by kgignatyev** [comment]
does not seem to be fixed in 2.5.2, or rather fix is unstable, when I run my suite it sometimes works, and sometimes write enum not as as string but as pair of class name and value :(

2.4.5 works reliably

**Comment by kgignatyev** [comment]
One thing that I noticed: with 2.5.2 it does not work every **second**-ish time with when I compile with using **external ZINC compile server**, when I do NOT use ZINC server it succeeds more often then it fails. 

  val  nm = new ObjectMapper() with ScalaObjectMapper
    nm.registerModule(DefaultScalaModule)
    println("n:\n"+nm.writeValueAsString(cp))

crazzy, I run 
mvn clean test -Dsuites=kgi.mcytravel.wsite.services.ContentPageServiceTest
and it prints output fine

{"contents":[{"l":"sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam"},{"r":{"contentId":"a","contentType":"P"}}],"created":null,"createdBy":"","lastUpdated":null,"lastUpdatedBy":"","ownerPointer":{"id":"2","entityType":"ml_"},"_id":"5c30540e-d2e7-4ab9-a2cc-9002804e573d"}

I run the same command again and now enum is not handled properly:
mvn clean test -Dsuites=kgi.mcytravel.wsite.services.ContentPageServiceTest

{"contents":[{"l":"ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat,"},{"r":{"contentId":"a","contentType":{"enumClass":"kgi.mcytravel.wsite.data.db.ContentType","value":"P"}}}],"created":null,"createdBy":"","lastUpdated":null,"lastUpdatedBy":"","ownerPointer":{"id":"2","entityType":"ml_"},"_id":"d1e176c4-be90-4932-a160-5566c1a1fb2f"}
